### PR TITLE
Improve APIcast CRD fields descriptions

### PR DESCRIPTION
**Pull request by miguelsorianod**

### CODE REVIEW

File: bundle/manifests/apps.3scale.net_apicasts.yaml [hunk 1]
```diff
@@ -38,13 +38,14 @@ spec:
             description: APIcastSpec defines the desired state of APIcast
             properties:
               adminPortalCredentialsRef:
-                description: LocalObjectReference contains enough information to let you locate the referenced object inside the same namespace.
+                description: Secret reference to a Kubernetes Secret containing the portal endpoint URL information. The Secret must be located in the same namespace
```
Comment (thomasmaas) [inline bundle/manifests/apps.3scale.net_apicasts.yaml:41]:
```suggestion
                description: Secret reference to a Kubernetes Secret containing the admin portal endpoint URL. The Secret must be located in the same namespace.
```

File: bundle/manifests/apps.3scale.net_apicasts.yaml [hunk 2]
```diff
@@ -38,13 +38,14 @@ spec:
...
                 properties:
                   name:
                     description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names TODO: Add other useful fields. apiVersion, kind, uid?'
```
Comment (thomasmaas) [inline bundle/manifests/apps.3scale.net_apicasts.yaml:44]: Seems you missed this one?

Comment (miguelsorianod) [reply bundle/manifests/apps.3scale.net_apicasts.yaml:44]: Unfortunately we can't control that one. That field `Name` is part of the `v1.LocalObjectReference` K8s type (see https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#localobjectreference-v1-core).
The description texts are taken from comments in the source code. The source code belongs to K8s so we don't have control over the descriptions shown for the attributes of the fields that are K8s objects.

File: bundle/manifests/apps.3scale.net_apicasts.yaml (hunk 1 above)
Comment (miguelsorianod) [reply bundle/manifests/apps.3scale.net_apicasts.yaml]: Updated :+1: .
I also added periods at the end of all descriptions

File: apis/apps/v1alpha1/apicast_types.go [hunk 3]
```diff
@@ -28,46 +28,81 @@ import (
...
 	Image *string `json:"image,omitempty"`
+	// ExposedHost is the domain name used for external access. By default no
+	// external access is configured.
 	// +optional
 	ExposedHost *APIcastExposedHost `json:"exposedHost,omitempty"`
+	// DeploymentEnvironment is the environment for which the configuration will
+	// be downloaded from 3scale (Staging or Production), when using new APIcast.
+	// The value will also be used in the header X-3scale-User-Agent in the
+	// authorize/report requests made to 3scale Service Management API. It is
+	// used by 3scale just for statistics.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:64]:
```suggestion
	// used by 3scale for statistics.
```

File: apis/apps/v1alpha1/apicast_types.go [hunk 4]
```diff
@@ -28,46 +28,81 @@ import (
...
+	// This setting should only be used for dev/testing purposes. Setting
+	// this disables automated upgrades of the image.
 	// +optional
 	Image *string `json:"image,omitempty"`
+	// ExposedHost is the domain name used for external access. By default no
+	// external access is configured.
 	// +optional
 	ExposedHost *APIcastExposedHost `json:"exposedHost,omitempty"`
+	// DeploymentEnvironment is the environment for which the configuration will
+	// be downloaded from 3scale (Staging or Production), when using new APIcast.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:61]:
```suggestion
	// be downloaded from 3scale (Staging or Production), when using APIcast.
```

File: apis/apps/v1alpha1/apicast_types.go [hunk 5]
```diff
@@ -28,46 +28,81 @@ import (
...
+	// LogLevel controls the log level of APIcast's OpenResty logs.
 	// +optional
 	// +kubebuilder:validation:Enum=debug;info;notice;warn;error;crit;alert;emerg
 	LogLevel *string `json:"logLevel,omitempty"` // APICAST_LOG_LEVEL
+	// PathRoutingEnabled can be used to enable APIcast's path-based routing
+	// in addition to to the default host-based routing.
 	// +optional
 	PathRoutingEnabled *bool `json:"pathRoutingEnabled,omitempty"` // APICAST_PATH_ROUTING
+	// ResponseCodesIncluded can be set to log the response codes of the responses
+	// returned by the API backend in 3scale.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:88]: Not sure what the "API backend in 3scale" refers to: 3scale API?

File: apis/apps/v1alpha1/apicast_types.go [hunk 6]
```diff
@@ -28,46 +28,81 @@ import (
...
 	// +optional
 	ResponseCodesIncluded *bool `json:"responseCodesIncluded,omitempty"` // APICAST_RESPONSE_CODES
+	// The period (in seconds) that the APIcast configuration will be stored in
+	// APIcast's cache.
 	// +optional
 	CacheConfigurationSeconds *int64 `json:"cacheConfigurationSeconds,omitempty"` // APICAST_CONFIGURATION_CACHE
+	// ManagementAPIScope controls APIcast Management API scope. The Management
+	// API is powerful and can control the APIcast configuration. debug level
+	// should only be enabled for debugging purposes.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:97]: Should we list possible values?

File: apis/apps/v1alpha1/apicast_types.go [hunk 7]
```diff
@@ -103,12 +138,13 @@ type APIcastSpec struct {
 	// HTTPSCertificateSecretRef references secret containing the X.509 certificate in the PEM format and the X.509 certificate secret key.
 	// +optional
 	HTTPSCertificateSecretRef *v1.LocalObjectReference `json:"httpsCertificateSecretRef,omitempty"`
+	// Workers defines the number of APIcast's worker processes per pod.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:141]: is this the correct description of `HTTPSCertificateSecretRef`?

File: apis/apps/v1alpha1/apicast_types.go [hunk 8]
```diff
@@ -155,7 +191,7 @@ type APIcastCondition struct {
 
 	// The Reason, Message, LastHeartbeatTime and LastTransitionTime fields are
 	// optional. Unless we really use them they should directly not be used even
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:193]: This reads a bit weird. Probably don't use we and consider explaining why they should not be used?

File: apis/apps/v1alpha1/apicast_types.go [hunk 9]
```diff
@@ -183,7 +219,7 @@ type APIcast struct {
 
 // +kubebuilder:object:root=true
 
-// APIcastList contains a list of APIcast
+// APIcastList contains a list of APIcast.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:222]:
```suggestion
// APIcastList contains a list of APIcasts.
```

File: apis/apps/v1alpha1/apicast_types.go (hunk 5 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go:88]: I'm not sure either. This description (and most of them) have been extracted from the source of truth which is APIcast's documentation. For this one specifically, it is https://github.com/3scale/APIcast/blob/master/doc/parameters.md#apicast_response_codes

File: apis/apps/v1alpha1/apicast_types.go (hunk 6 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go:97]: A little bit below you can find the values validation performet. Additionally, lots of the CR parameters belong to the APIcast realm  so user should already have this knowledge / consult it into the APIcast doc.

File: apis/apps/v1alpha1/apicast_types.go (hunk 5 above)
Comment (thomasmaas) [reply apis/apps/v1alpha1/apicast_types.go:88]: @eloycoto do you know what is meant exactly?

File: apis/apps/v1alpha1/apicast_types.go (hunk 8 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go:193]: This is not shown on the generated CRD files. It is purely a code comment.

File: apis/apps/v1alpha1/apicast_types.go (hunk 9 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go]: fixed

File: apis/apps/v1alpha1/apicast_types.go (hunk 4 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go]: fixed

File: apis/apps/v1alpha1/apicast_types.go (hunk 3 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go]: fixed

File: apis/apps/v1alpha1/apicast_types.go (hunk 5 above)
Comment (thomasmaas) [reply apis/apps/v1alpha1/apicast_types.go:88]: or maybe @davidor?

Comment (davidor) [reply apis/apps/v1alpha1/apicast_types.go:88]: When that is enabled, APIcast sends the response code returned by the upstream API to apisonator, so you can see those response codes in the system UI.

File: apis/apps/v1alpha1/apicast_types.go [hunk 10]
```diff
@@ -28,46 +28,81 @@ import (
...
 	ConfigurationLoadMode *string `json:"configurationLoadMode,omitempty"` // APICAST_CONFIGURATION_LOADER
+	// LogLevel controls the log level of APIcast's OpenResty logs.
 	// +optional
 	// +kubebuilder:validation:Enum=debug;info;notice;warn;error;crit;alert;emerg
 	LogLevel *string `json:"logLevel,omitempty"` // APICAST_LOG_LEVEL
+	// PathRoutingEnabled can be used to enable APIcast's path-based routing
+	// in addition to to the default host-based routing.
 	// +optional
 	PathRoutingEnabled *bool `json:"pathRoutingEnabled,omitempty"` // APICAST_PATH_ROUTING
+	// ResponseCodesIncluded can be set to log the response codes of the responses
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:87]:
```suggestion
	// ResponseCodesIncluded when set to true, stores the response codes that the managed API returns
```

File: apis/apps/v1alpha1/apicast_types.go [hunk 11]
```diff
@@ -28,46 +28,81 @@ import (
...
+	// returned by the API backend in 3scale.
```
Comment (thomasmaas) [inline apis/apps/v1alpha1/apicast_types.go:88]:
```suggestion
	// in Apisonator, so they can then be visualized in the 3scale admin portal.
```

File: apis/apps/v1alpha1/apicast_types.go (hunk 10 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go:87]: I'm not sure about this one. It does not store anything right? My understanding was that it just includes the response codes as part of the log :thinking:

Comment (eguzki) [reply apis/apps/v1alpha1/apicast_types.go:87]: https://github.com/3scale/APIcast/blob/master/doc/parameters.md#apicast_response_codes

Comment (eguzki) [reply apis/apps/v1alpha1/apicast_types.go:87]: It basically starts counting response codes per application and then reports to 3scale backend (apisonator). The that data is stored (by backend listener) in redis and available in the dashboard

Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go:87]: I think the description that's currently shown is good enough. No need to go into that detail about redis storage

Comment (eguzki) [reply apis/apps/v1alpha1/apicast_types.go:87]: I agree. My comment was to let you both know what this parameter activates.

File: apis/apps/v1alpha1/apicast_types.go (hunk 11 above)
Comment (miguelsorianod) [reply apis/apps/v1alpha1/apicast_types.go]: Updated

**Comment by miguelsorianod** [comment]
@thomasmaas , @eguzki is this ready to be merged?

File: apis/apps/v1alpha1/apicast_types.go (hunk 11 above)
Comment (eguzki) [reply apis/apps/v1alpha1/apicast_types.go]: Apisonator might no be known in this context ("apicast operator repo"). I would say:

`ResponseCodesIncluded can be set to log the response codes of the responses in the 3scale backend storage, so they can then be visualized in the 3scale admin portal.`

**Review by thomasmaas — APPROVED** [review]

**Pull request merged**